ecmascript/transforms/typescript/src/strip.rs

```diff
@@ -1113,26 +1113,50 @@ impl Strip {
 impl Visit for Strip {
     fn visit_ident(&mut self, n: &Ident, _: &dyn Node) {
         let is_type_only_export = self.is_type_only_export;
-        self.scope
-            .imported_idents
+        let entry = self
+            .scope
+            .referenced_idents
             .entry((n.sym.clone(), n.span.ctxt()))
-            .and_modify(|v| v.has_concrete |= !is_type_only_export);
+            .or_default();
+        entry.has_concrete |= !is_type_only_export;
 
         n.visit_children_with(self);
     }
 
-    fn visit_class_decl(&mut self, n: &ClassDecl, _: &dyn Node) {
-        n.visit_children_with(self);
-
-        self.decl_names.insert(n.ident.to_id());
-    }
-
     fn visit_decl(&mut self, n: &Decl, _: &dyn Node) {
         self.handle_decl(n);
 
         let old = self.non_top_level;
         self.non_top_level = true;
-        n.visit_children_with(self);
+        // Visit the declaration body and not the identifier. Otherwise
+        // type declarations could be unintentionally marked as concrete
+        // by the visit_ident function above.
+        match n {
+            Decl::Class(class) => {
+                self.decl_names.insert(class.ident.to_id());
+                class.class.visit_with(class, self);
+            }
+            Decl::Fn(f) => f.function.visit_with(f, self),
+            Decl::Var(ref var) => {
+                for decl in &var.decls {
+                    decl.init.visit_with(decl, self);
```

Comment:
This can break code if it contains a default value for a pattern.

For example, 

```ts
import { b } from './foo';
const [a = b] = []
```

`b` will not be visited.

Reply:
I typically workaround this by storing a flag (in_var_pat or something) in the struct and ignoring `Pat::Ident` if the flag is set.

Reply:
Note that you have to visit children nodes of other patterns, because they may contain some expressions as a default value.

Reply:
Good catch.